retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/ArmeriaRetrofitBuilder.java

```diff
@@ -41,127 +39,118 @@
+        baseWebClientHost = baseUrl.host();
+        baseWebClientPort = baseUrl.port();
 
-    /**
-     * Sets the API base URL.
-     *
-     * @see Builder#baseUrl(String)
-     */
-    public ArmeriaRetrofitBuilder baseUrl(URI baseUrl) {
-        requireNonNull(baseUrl, "baseUrl");
-        checkArgument(SessionProtocol.find(baseUrl.getScheme()) != null,
-                      "baseUrl must have an HTTP scheme: %s", baseUrl);
-        final String path = baseUrl.getPath();
-        if (!path.isEmpty() && !SLASH.equals(path.substring(path.length() - 1))) {
-            throw new IllegalArgumentException("baseUrl must end with /: " + baseUrl);
-        }
-        this.baseUrl = baseUrl.toString();
-        return this;
-    }
+        // Re-create the base client without a path, because Retrofit will always provide a full path.
+        baseWebClient = WebClient.builder(protocol,
+                                          webClient.endpointGroup())
+                                 .factory(webClient.factory())
+                                 .options(webClient.options())
+                                 .build();
 
-    /**
-     * Sets the {@link ClientOptions} that customizes the underlying {@link WebClient}.
-     * This method can be useful if you already have an Armeria client and want to reuse its configuration,
-     * such as using the same decorators.
-     * <pre>{@code
-     * WebClient myClient = ...;
-     * // Use the same settings and decorators with `myClient` when sending requests.
-     * builder.clientOptions(myClient.options());
-     * }</pre>
-     */
-    public ArmeriaRetrofitBuilder clientOptions(ClientOptions clientOptions) {
-        requireNonNull(clientOptions, "clientOptions");
-        return withClientOptions((uri, b) -> b.options(clientOptions));
+        nonBaseClientFactory = (p, url) -> WebClient.builder(p, Endpoint.of(url.host(), url.port()))
+                                                    .factory(baseWebClient.factory())
+                                                    .options(baseWebClient.options())
+                                                    .build();
     }
 
     /**
-     * Sets the {@link BiFunction} that customizes the underlying {@link WebClient}.
+     * Specifies the {@link BiFunction} that creates a new non-base {@link WebClient}, which is used for
+     * sending requests to other authorities than that of base URL. If not specified, the non-base
+     * {@link WebClient} will have the same options with the base {@link WebClient}, which was specified
+     * with {@link ArmeriaRetrofit#of(WebClient)} or {@link ArmeriaRetrofit#builder(WebClient)}.
+     *
+     * <p>To avoid the overhead of repetitive instantiation of {@link WebClient}s, the {@link WebClient}s
+     * returned by the specified {@link BiFunction} will be cached for each combination of:
+     * <ul>
+     *   <li>Whether the connection is secured (HTTPS or HTTPS)</li>
```

Comment:
If this is the cache key, I guess it should only be these three fields that are passed to the callback, not `HttpUrl` - it's possible to create a broken callback if inspecting other fields of `HttpUrl`.

Reply:
I didn't want to bother adding a factory interface, so... :laughing: Let me try to clean it up.

Reply:
Just fixed by passing `Endpoint` instead of `HttpUrl`.